core/shared/platform/windows/win_file.c

```diff
@@ -4,117 +4,766 @@
+    else
+        file_info_class = FileIdBothDirectoryInfo;
+
+    if (dir_stream->cursor == 0
+        && !GetFileInformationByHandleEx(dir_stream->handle->raw.handle,
+                                         file_info_class, dir_stream->info_buf,
+                                         sizeof(dir_stream->info_buf))) {
+        if (out_entry != NULL)
+            *out_entry = NULL;
+        DWORD win_error = GetLastError();
+        // We've reached the end of the directory - return success
+        if (win_error == ERROR_NO_MORE_FILES) {
+            dir_stream->cookie = 0;
+            dir_stream->cursor = 0;
+            return __WASI_ESUCCESS;
+        }
+
+        return convert_windows_error_code(win_error);
+    }
+
+    FILE_ID_BOTH_DIR_INFO *current_info =
+        (FILE_ID_BOTH_DIR_INFO *)(dir_stream->info_buf + dir_stream->cursor);
+
+    if (current_info->NextEntryOffset == 0)
+        dir_stream->cursor = 0;
+    else
+        dir_stream->cursor += current_info->NextEntryOffset;
+
+    ++dir_stream->cookie;
+
+    if (out_entry != NULL)
+        *out_entry = current_info;
+    else
+        return __WASI_ESUCCESS;
+
+    // Convert and copy over the wchar filename into the entry_name buf
+    int ret = WideCharToMultiByte(
+        CP_UTF8, 0, current_info->FileName,
+        current_info->FileNameLength / (sizeof(wchar_t) / sizeof(char)),
+        dir_stream->current_entry_name, sizeof(dir_stream->current_entry_name),
+        NULL, NULL);
+
+    if (ret == 0)
+        return convert_windows_error_code(GetLastError());
+
+    return __WASI_ESUCCESS;
+}
+
+static HANDLE
+create_handle(wchar_t *path, bool is_dir, bool follow_symlink, bool readonly)
+{
+    CREATEFILE2_EXTENDED_PARAMETERS create_params;
+
+    create_params.dwSize = sizeof(create_params);
+    create_params.dwFileAttributes = FILE_ATTRIBUTE_NORMAL;
+    create_params.dwSecurityQosFlags = 0;
+    create_params.dwFileFlags = 0;
+    create_params.lpSecurityAttributes = NULL;
+    create_params.hTemplateFile = NULL;
+
+    if (is_dir) {
+        create_params.dwFileAttributes |= FILE_ATTRIBUTE_DIRECTORY;
+        create_params.dwFileFlags |= FILE_FLAG_BACKUP_SEMANTICS;
+    }
+
+    if (!follow_symlink)
+        create_params.dwFileFlags |= FILE_FLAG_OPEN_REPARSE_POINT;
+
+    DWORD desired_access = GENERIC_READ;
+
+    if (!readonly) {
+        desired_access |= GENERIC_WRITE;
+        create_params.dwFileAttributes |= FILE_ATTRIBUTE_READONLY;
+    }
+
+    return CreateFile2(path, desired_access,
+                       FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                       OPEN_EXISTING, &create_params);
+}
+
+#if WINAPI_PARTITION_DESKTOP
+// Modifies the given path in place and replaces it with the filename component
+// (including the extension) of the path.
+static __wasi_errno_t
+extract_filename_from_path(wchar_t *path, size_t buf_size)
+{
+    wchar_t extension[256];
+    wchar_t filename[256];
```

Comment:
Buffer size of filename may be a little small?

Reply:
Going by https://learn.microsoft.com/en-gb/windows/win32/fileio/maximum-file-path-limitation?tabs=registry, the max filename length is 256 characters.

Reply:
Got it, thanks.